primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,130 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use crate::v2::{BlakeTwo256, HashT as _};
+use parity_scale_codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use parity_util_mem::MallocSizeOf;
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec, vec::Vec};
+
+/// # Execution environment parameter tags
+/// Values from 1 to 16 are reserved for system-wide parameters not related to any concrete
+/// execution environment.
+/// Environment type
+pub const EEPAR_01_ENVIRONMENT: u32 = 1;
+/// Logical stack limit, in stack items
+pub const EEPAR_17_STACK_LOGICAL_MAX: u32 = 17;
+/// Native stack limit, in bytes
+pub const EEPAR_18_STACK_NATIVE_MAX: u32 = 18;
+
+/// # Execution enviroment types
+/// Generic Wasmtime environment
+pub const EXEC_ENV_TYPE_WASMTIME_GENERIC: u32 = 0;
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+/// # Deterministically serialized execution environment semantics
+/// Represents an arbitrary semantics of an arbitrary execution environment, so should be kept as
+/// abstract as possible. Mapping from `u32` constant tags to SCALE-encoded values was chosen
+/// over an idiomatic `enum` representation as the latter would require caller to rely on
+/// deserialization errors when decoding a structure of newer version not yet known to the node
+/// software, which is undesirable.
+/// As this one goes on-chain, it requires full serialization determinism. That is achieved by
+/// enforcing constant tags to be in ascending order and using deterministic SCALE-codec for values.
+#[cfg_attr(feature = "std", derive(MallocSizeOf))]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub struct ExecutorParams(Vec<(u32, Vec<u8>)>);
```

Comment:
This looks like it should be a map/dictionary?

Reply:
Not sure if this answers the question, but there's some more information at this comment: https://github.com/paritytech/polkadot/pull/6161#discussion_r1001195034

Reply:
@eskimor and [it indeed was](https://github.com/paritytech/polkadot/pull/6161/commits/a3ab2512b301290113addb6173638af83949809f#diff-59bd1090a32a865b4059c38707d7abe125a65fa65d955abd0df38a1ba6d3c345R113) at some point of time.

Vector of `u32 -> Vec<u8>` mapping was agreed with @pepyakin at the very beginning, I implemented it and it looked ugly to me, so I re-implemented it in an idiomatic way, but after some more discussions with @pepyakin I reverted to the abstract mapping.